### PR TITLE
style: add jaxtyping annotations and contributor documentation

### LINKED ISSUE #353: [FEATURE] Integrate torchtyping / jaxtyping for tensor shape & type annotations
## Is your feature request related to a problem? Please describe.
Our codebase uses PyTorch extensively, but we currently lack a consistent way to annotate tensor shapes and dtypes. [torchtyping](https://github.com/patrick-kidger/torchtyping) (now `jaxtyping`) extends Python type hints with shape and dtype information for tensors, improving readability and catching errors earlier during development.

This issue proposes integrating `torchtyping`/`jaxtyping` into the project.

## Describe the solution you'd like
- [ ] Add torchtyping as a dependency (likely a dev dependency, but contributors can discuss whether runtime use is desired).
- [ ] Update one or two representative modules (e.g., model definitions or data preprocessing) to demonstrate how to use torchtyping annotations.
- [ ] Add tests to verify that torchtyping integration plays well with our type checker.
- [ ] Update contributor documentation with a short section on how/why we use torchtyping.

## Additional context
Example:
```python
from torchtyping import TensorType
import torch

Batch = TensorType["batch"]
Features = TensorType["features"]

def forward(x: TensorType[Batch, Features, torch.float32]) -> TensorType[Batch, Features, torch.float32]:
    ...
```

### PR DESCRIPTION
**Pull request by vaishnaviparabkar90**
## Description

This PR adds jaxtyping (torchtyping) annotations to key collate functions
(image_classification_collate) to improve code readability and early error detection.

Additionally, a new contributor documentation section has been added in the
user manual to explain tensor annotations, usage examples, and optional runtime
validation with check_type.

- Annotated image_classification_collate function with Float and Int tensor types
- Added runtime shape/dtype checks (optional)
- Added `user_manual/tensor_annotations.rst` with purpose, examples, and best practices

## Related Issue

Fixes #353

---

> [!NOTE]
> Adds jaxtyping guidance and dependency, clarifies image classification collate returns, and expands tests to assert batch shapes/dtypes.
> 
> - **Documentation**:
>   - Add "Tensor Annotations with jaxtyping" section in `docs/contributions/how_to_contribute.rst`.
>   - Clarify `image_classification_collate` return shapes/dtypes in its docstring.
> - **Dependencies**:
>   - Add `jaxtyping` to `pyproject.toml` project dependencies.
> - **Tests**:
>   - Strengthen `tests/data/test_datamodule.py` by materializing a train batch and asserting `images` channel count/size and dtypes (`float32` images, `int64` labels).
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 24769a92ce52ee074ca7c12da8bb7be338863b32. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: docs/user_manual/tensor_annotations.rst [hunk 1]
```diff
@@ -0,0 +1,94 @@
+Tensor Annotations with jaxtyping
```
Comment (johannaSommer) [inline docs/user_manual/tensor_annotations.rst]: Thanks a lot for already adding this to the documentation! Could you please shorten this into e.g. only one paragraph (which can then link to relevant documentation from jaxtyping)? I would also suggest to add this paragraph to [this](https://github.com/PrunaAI/pruna/blob/main/docs/contributions/how_to_contribute.rst) section of the documentation, as it's mainly relevant to contributors :)

**Review by johannaSommer — CHANGES_REQUESTED** [review]
Thank you so much @vaishnaviparabkar90 looks great! Just have one quick comment on the documentation, otherwise I will already run the workflows and we are good to go 👏🏻

File: docs/user_manual/tensor_annotations.rst (hunk 1 above)
Comment (vaishnaviparabkar90) [reply docs/user_manual/tensor_annotations.rst]: Changes applied after review.

Comment (sdiazlor) [reply docs/user_manual/tensor_annotations.rst]: Thanks @vaishnaviparabkar90 . In fact, could you delete this rst and leave it in how_to_contribute?

Comment (vaishnaviparabkar90) [reply docs/user_manual/tensor_annotations.rst]: ok

File: src/pruna/data/collate.py [hunk 2]
```diff
@@ -137,23 +137,10 @@ def image_classification_collate(
     """
     Custom collation function for image classification datasets.
 
-    Expects a ``image`` column containing PIL images and a ``label`` column containing the class label in the dataset.
-
-    Parameters
```
Comment (sdiazlor) [inline src/pruna/data/collate.py:142]: Hi! Is there a reason for this docstring deletion?
